### PR TITLE
Recursive Test Discovery [V3]

### PR DESCRIPTION
**Pull request by apahim**
v3:
- Remove commit to order test classes.
- Add parents methods to the child class (instead of add parents classes with their methods).
- In consequence of the previous item, remove the commit adding the path information to the internal 'info' structure.
- Squash the 2 remaining commits since they are related to the same functionality.

v2: #2053 
- `:avocado: disable` works as usual, and in this v2 it has no effect in a recursive discovery.
- Improved docs, addressing v1 comments.

v1: #2052 
- Recursion, when engaged (using `:avocado: recursive`), will discover all parents up to the base class.
- `:avocado: enable` works as usual, but it's not needed in a recursive discovery.
- `:avocado: disable` works as usual, but will abort the recursive discovery from that point.

### CODE REVIEW

File: avocado/core/loader.py [hunk 1]
```diff
@@ -615,6 +617,9 @@ def _find_avocado_tests(self, path):
         # The resulting test classes
         result = {}
 
+        if os.path.isdir(path):
+            path = os.path.join(path, "__init__.py")
```
Comment (ldoktor) [inline avocado/core/loader.py]: not worth changing it now, but if it comes to sending next version, you should split this change into a separate commit (it adds different functionality)

File: docs/source/WritingTests.rst [hunk 2]
```diff
@@ -1251,6 +1251,68 @@ The docstring ``:avocado: disable`` is evaluated first by Avocado,
...
+        """
+
+        def test_second_child(self):
+            pass
+
+Using only `test_second_child.py` as a test reference will result in::
+
+    $ avocado list test_second_child.py
+    INSTRUMENTED test_second_child.py:SecondChild.test_second_child
+    INSTRUMENTED test_second_child.py:FirstChild.test_first_child
```
Comment (ldoktor) [inline docs/source/WritingTests.rst]: Actually this behavior changed and it'll always be `test_second_child.py:SecondChild:` followed by the 3 methods.

File: docs/source/WritingTests.rst [hunk 3]
```diff
@@ -1251,6 +1251,68 @@ The docstring ``:avocado: disable`` is evaluated first by Avocado,
...
+    INSTRUMENTED test_second_child.py:BaseClass.test_basic
+
+Notice that the other docstrings will be ignored in ancestors during the
+recursive discovery. What means that even if an ancestor contains the
+docstring `:avocado: disable`, that ancestor will still be included in
```
Comment (ldoktor) [inline docs/source/WritingTests.rst]: Perhaps you could mention that in case of `:avocado: recursive` as well as with `:avocado: enable` avocado does not check whether the test is inherited from `avocado.Test`, so it is possible to force-execute non-avocado tests as avocado ones (where the user is responsible for creating the necessary methods). Anyway it's just a suggestion and I'm not entirely sure we should go into that much details.

File: avocado/core/loader.py (hunk 1 above)
Comment (apahim) [inline avocado/core/loader.py]: sure

File: docs/source/WritingTests.rst (hunk 2 above)
Comment (apahim) [inline docs/source/WritingTests.rst]: oh, good catch

**Comment by apahim** [comment]
On it right now.

File: docs/source/WritingTests.rst (hunk 3 above)
Comment (apahim) [inline docs/source/WritingTests.rst]: Let's keep that in mind, but not include in the docs.
